dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -302,243 +297,294 @@ Block DMFileReader::read()
+                }
+            }
+        }
+    }
+    else if (enable_handle_clean_read)
+    {
+        handle_column_clean_read_packs.reserve(read_packs);
+        version_column_clean_read_packs.reserve(read_packs);
+        del_column_clean_read_packs.reserve(read_packs);
+        for (size_t i = start_pack_id; i < next_pack_id; ++i)
+        {
+            // If all handle in a pack are in the given range, no not_clean rows, and max version <= max_read_version,
+            // we do not need to read handle column.
+            if (handle_res[i] == All && pack_stats[i].not_clean == 0
+                && pack_filter.getMaxVersion(i) <= max_read_version)
+            {
+                handle_column_clean_read_packs.insert(i);
+                version_column_clean_read_packs.insert(i);
+                del_column_clean_read_packs.insert(i);
+            }
+        }
+    }
 
-    size_t read_rows = 0;
-    size_t not_clean_rows = 0;
-    size_t deleted_rows = 0;
+    /// 4. Read columns.
 
-    const std::vector<RSResult> & handle_res = pack_filter.getHandleRes(); // alias of handle_res in pack_filter
-    RSResult expected_handle_res = handle_res[next_pack_id];
-    for (; next_pack_id < use_packs.size() && use_packs[next_pack_id] && read_rows < rows_threshold_per_read;
-         ++next_pack_id)
+    ColumnsWithTypeAndName columns;
+    columns.reserve(read_columns.size());
+    for (size_t i = 0; i < read_columns.size(); ++i)
     {
-        if (read_pack_limit != 0 && next_pack_id - start_pack_id >= read_pack_limit)
-            break;
-        if (enable_handle_clean_read && handle_res[next_pack_id] != expected_handle_res)
-            break;
-
-        read_rows += pack_stats[next_pack_id].rows;
-        not_clean_rows += pack_stats[next_pack_id].not_clean;
-        // Because deleted_rows is a new field in pack_properties, we need to check whehter this pack has this field.
-        // If this pack doesn't have this field, then we can't know whether this pack contains deleted rows.
-        // Thus we just deleted_rows += 1, to make sure we will not do the optimization with del column(just to make deleted_rows != 0).
-        if (static_cast<size_t>(pack_properties.property_size()) > next_pack_id
-            && pack_properties.property(next_pack_id).has_deleted_rows())
+        auto & cd = read_columns[i];
+        try
         {
-            deleted_rows += pack_properties.property(next_pack_id).deleted_rows();
+            ColumnPtr col;
+            // For handle, tag and version column, we can try to do clean read.
+            switch (cd.id)
+            {
+            case EXTRA_HANDLE_COLUMN_ID:
+                col = readExtraColumn(cd, start_pack_id, read_packs, read_rows, handle_column_clean_read_packs, i);
+                break;
+            case TAG_COLUMN_ID:
+                col = readExtraColumn(cd, start_pack_id, read_packs, read_rows, handle_column_clean_read_packs, i);
```

Comment:
Is `handle_column_clean_read_packs` for `TAG_COLUMN_ID` a mistake here?

Reply:
yes, fixed...